updates documentation for builder dev environments with officially su…

nellshamrell opened the pull request:
…pported path

I want to start off this pull request with yes, we should automate this process and harness the habitat studio to do so.  That has not been done yet, though, and there is enormous pain associated with setting up a Builder Dev environment in the meantime.  This documentation is meant to be a stop gap until we have a better dev environment setup - it provides one, officially supported path that definitely works right now.  People are free to use other setups at their own risk, but this provides a "happy path" - one way which will work.

There are definitely more improvements we can make to this process - I think some of the scripts need some additions - but again, this documentation is supposed to capture the state of setting up a builder dev environment right now.

When #4161 is complete - which provides a vagrant setup - we should discuss whether we want to officially support that workflow, or include it as another "at your own risk" workflow, I am open to either.

Signed-off-by: Nell Shamrell <nellshamrell@gmail.com>

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

nellshamrell commented:
These instructions have been verified on a VMWare Fusion VM running Ubuntu 17.10

rsertelon commented:
I will try it on Ubuntu 16.04 LTS and virtualbox. If I see problems I'll tell you. Thank you for this documentation upgrade :)

rsertelon commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,285 @@
 
-When `make bldr-run` can proceed with all the services up and running, you are ready to proceed to the next step.
+* Navigate to http://localhost:3000/#/pkgs
+* Log in
```
Following the instructions, I cannot login here. It may be caused by the missing `.secrets/builder-github-app.pem` file though, see previous comment about that :)

nellshamrell replied:
That is likely why, I will add a note to the documentation saying to expect info on being able to setup Github auth without a secret from the Habitat 1Password Vault soon.

rsertelon commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,285 @@
+```
+
+* Install the [Habitat CLI](https://www.habitat.sh/docs/get-habitat/)
+* Copy habitat-builder-dev.2017-10-02.private-key.pem from 1Password into path/to/habitat/repo/.secrets/builder-github-app.pem
````
I understand this guide is mainly for core contributors. For those who haven't access to secrets (like me ;)). How can we have our own secret, does that mean we'd need to fork the github app? (is it even possible?)

nellshamrell replied:
There is some work in progress by our partners at Kinvolk on this - will definitely update the docs when it is ready to be merged https://github.com/habitat-sh/habitat/pull/4161

schu commented:
Hey, thanks a lot for working on the dev setup :)

My main concerns with the document as-is are that

* it cannot work for external folks (Habitat-internal secrets used),
* it doesn't explain configuration in detail (hard to setup correctly, especially GitHub App and auth; not possible w/o changing tracked code),
* it is a fairly long and manual setup where IMHO the setup should be as automated as possible (to not only be easily used but also tested and maintained) and
* it seems to give instructions which are not actually needed today, for example setting up `HAB_AUTH_TOKEN` while the repository also maintains a [`.envrc`](https://github.com/habitat-sh/habitat/blob/master/.envrc) file.

With the guide & setup in https://github.com/habitat-sh/habitat/pull/4161 we have automated most of the manual work (after setting up configuration) so that only a [few commands](https://github.com/kinvolk/habitat/blob/schu/vagrant/BUILDER_VAGRANT.md#setup) are actually needed to a running, local builder instance. And we were able to do that with only a few changes / additions, i.e. a lot of automation is already there but not used and/or documented.

Personally I think the canonical "officially supported" setup should be a fully automated environment (modulo upfront configuration that is clearly explained).

nellshamrell commented:
@schu Thank you for the comments!  I completely agree the long term solution should be an automated setup and when your pull request is ready to be merged we should consider that one.  This is meant only to provide a manual stop gap until an automated solution is ready.  

Right now, the setup is immensely painful for both core maintainers and non-core maintainers.  I realize we do need to work on focusing on those who do not have access to Habitat secrets - and your work is fantastic on this, and I'd love to adopt it as soon as it is ready! - but, again, this is meant as a stop gap until your solution is ready to merge.

nellshamrell commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,285 @@
+* All of the following steps should be run within your Virtual Machine
+* A [Github Personal Access token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/) with all repo and all user permissions.
+* The [Habitat Builder App](https://github.com/apps/habitat-builder) installed on your Github account.
+* Open your Sudoers file at /etc/sudoers and remove the "secure_path" section (this is required for the make scripts to run correctly)
```
Good point.  Doing so.

nellshamrell commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,285 @@
+
+* Install git
+```
+$ sudo apt-get install git
````
Added.

baumanj replied:
`sudo apt-get -y --no-install-recommends` could save a bit of time. I could really go either way.

nellshamrell commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,285 @@
-backend = "local"
-local_dir = "/tmp"
+```
+$ sudo ln -s ~/habitat /src
````
Fixing.

nellshamrell commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,285 @@
+
+* Add your Github Auth Key:
+
+**/hab/svc/builder-worker**
```
Good catch!  Fixing.

nellshamrell commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,285 @@
-sessionsrv: target/debug/bldr-sessionsrv start --config /home/your_alias/habitat/config_sessionsrv.toml
-worker: target/debug/bldr-worker start --config /home/your_alias/habitat/config_worker.toml
-jobsrv: target/debug/bldr-jobsrv start --config /home/your_alias/habitat/config_jobsrv.toml
+$ npm install
```
Good point.  I think in the case of this guide - providing a guaranteed to work workflow until we have something better - I'd like to only officially support 17.10 for the sake of consistency.

nellshamrell commented:
@rsertelon I believe I have addressed all of your comments.

I want to emphasize that although this is not a long term solution, it is meant only to be better than what we have right now - it has taken over a week for the last two core contributors to get a functioning environment with the current documentation, this improves that.  I'm looking forward to the automated solution in #4161 and an eventual Habitat studio development environment.

schu commented:
> I completely agree the long term solution should be an automated setup and when your pull request is ready to be merged we should consider that one. This is meant only to provide a manual stop gap until an automated solution is ready.

Ok :) Could you maybe add a note at the top of the document that makes it clear that it's only meant for Habitat-internal folks and a note to the other dev documents that they are unsupported and deprecated?

> Right now, the setup is immensely painful for both core maintainers and non-core maintainers. I realize we do need to work on focusing on those who do not have access to Habitat secrets - and your work is fantastic on this, and I'd love to adopt it as soon as it is ready! - but, again, this is meant as a stop gap until your solution is ready to merge.

The canonical, automated setup will only progress if the Habitat team contributes to it and, as important, commits to its usage. The updated Vagrant setup can be used and improved today, so I feel putting more work into manual setups goes in the wrong direction :)

schu commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
+$ sudo make build-srv
+```
+
+* Open up the builder-worker file
````
I think `auth_token` should be already set after https://github.com/habitat-sh/habitat/blob/master/support/builder/dev-config.sh#L348 is run (called from https://github.com/habitat-sh/habitat/blob/master/support/linux/provision.sh#L35).

Quickly testing in a clean machine seems to confirm that `auth_token` is indeed set correctly as long as `export HAB_AUTH_TOKEN="<your github token>"` was run before.

I.e. can this part be removed?

nellshamrell replied:
it does not work when I am testing, for some reason - even though I set HAB_AUTH_TOKEN beforehand.

mwrock replied:
maybe with `sudo -E` ?

nellshamrell replied:
I did set the sudo -E, and it still did not appear to work for me - but @baumanj has offered to run through these instructions again, if he has the same experience I will edit the instructions :)

nellshamrell replied:
oh, wait, I'm sorry @mwrock - I misread your post above.  I will try that and see if using sudo -E make build-srv fixes the issue.

baumanj replied:
There's a bit of a tricky situation here. If you export `HAB_BLDR_URL` _before_ running `provision.sh`, it will fail when it tries to run the [`hab pkg install` commands](https://github.com/habitat-sh/habitat/blob/master/support/linux/provision.sh#L23-L26). However, if you wait until after running `provision.sh` to add all the `HAB_` environment variables, you have to manually set the `auth_token` in `/hab/svc/builder-worker/config.toml`.

I see a couple options:
1. Set the `HAB_` env vars first and modify `provision.sh` to do its `hab pkg install`s with the `-u https://bldr.habitat.sh` option.
2. Set the `HAB_` env vars after `provision.sh` and set `auth_token` in `/hab/svc/builder-worker/config.toml` manually
3. Make `provision.sh` take the GitHub token as input and append the `HAB_` env var export lines to `~/.profile` _after_ the `hab pkg install`s.

Thoughts?

chefsalim replied:
Can we put the sudo inside the Makefile, so we don't need to type in a sudo -E ... every time to start up the env.

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
+* Run this command (it may take awhile) to start all the Builder services
+
+```
+$ sudo -E make bldr-run
````
This can be `bldr-run-no-build` since we just did the build step.

chefsalim replied:
We should also add a comment to make sure the services all start properly - wait till the end when you can see the worker heartbeat debug entries in the log.  Also as a troubleshooting step, if the services don't start because some were already running from a previous attempt, there is now a 'make bldr-kill' command that may be useful.

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
-build_worker_teams = [2555389]
-early_access_teams = [1995301]
 ```
+$ sudo make build-srv
````
I don't think this command should need `sudo`. I initially ran into this error:
```
jbauman@ubuntu:~/habitat$ make build-srv
sh -c 'cd components/builder-api && cargo build'
error: failed to load source for a dependency on `urlencoded`

Caused by:
  Unable to update https://github.com/iron/urlencoded#8157212e

Caused by:
  failed to open: /home/jbauman/.cargo/git/.cargo-lock-git

Caused by:
  Permission denied (os error 13)
Makefile:230: recipe for target 'build-builder-api' failed
make: *** [build-builder-api] Error 101
```

But that was just because the `~/.cargo` dir was owned by root, and I don't think we want that. It can be fixed by running 

`sudo chown -R jbauman:jbauman /home/jbauman/.cargo/`
or more generalized
`sudo USER=$(id -un) GROUP=$(id -gn) chown -R $USER:$GROUP /home/jbauman/.cargo/`

But I'd prefer to see if I can figure out a good way to have cargo installed by the user rather than root. That will probably involve changing the [`install_dev_9_linux.sh`](https://github.com/habitat-sh/habitat/blob/master/support/linux/install_dev_9_linux.sh) script that `provision.sh` calls.

baumanj replied:
On second thought, I think it would be better to run `provision.sh` without `sudo`. `install_dev_9_linux.sh` doesn't presume `sudo` since it adds `sudo` only to certain commands. I'm going to see if I can do the same for `provision.sh` so we're only using `sudo` where we really need it.

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
-backend = "local"
-local_dir = "/tmp"
 ```
+$ cd path/to/your/habitat/repo
````
We make the symlink, so we might as well use it. All the `path/to/your/habitat/repo`s can be replaced with `/src`, right?

baumanj commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,294 @@
-(Note: If you want your log files to persist across restarts of your development machine, replace `/tmp` with some other directory. It *must* exist and be writable before you start the job server).
+* Navigate to http://localhost:3000/#/pkgs
+* Log in
+* Click on "Create New Origin"
```
Should this be "Navigate to http://localhost:3000/#/origins" and "Create origin"? That's what my UI looks like, but maybe I'm doing something different?

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
+* Hit Ctrl + D twice
+* Run
+```
+$ hab install core/hab-backline
````
I had to run this with `sudo`

chefsalim replied:
yeah, hab install requires a sudo..

baumanj commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,294 @@
-There are a couple of ways to do this.
+### Create a Project
+* In UI
+* Click on the Origin your created in "Creating an Origin"
```
`s/your/you/`, I think

baumanj commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,294 @@
+* In UI
+* Click on the Origin your created in "Creating an Origin"
+* Click on "Connect a plan file"
+* Add in a github repo with a plan.sh file in it (I use "nellshamrell/sample-node-app")
```
Maybe change this to `habitat-sh/sample-node-app`? I can't access `nellshamrell/sample-node-app`.

chefsalim replied:
Yeah the access depends on the Github App - which is installed in the user's account, so it will only be able to see repos that the user grants it access to.  It may be better to use a core plans app (eg, core/nginx) as an example.

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
+* Open your Sudoers file at /etc/sudoers and remove the "secure_path" section (this is required for the make scripts to run correctly)
+
+```
+$ sudo visudo -f /etc/sudoers
````
For those of us who aren't `vi` users, it might be nice to supply an alternative. For me (since emacs isn't already installed) this works well:
```
$ EDITOR=$(which nano) sudoedit /etc/sudoers
```
Or it can be totally automated:
```
$ sudo sed -i.bak '/secure_path/s/^/#/' /etc/sudoers
```

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
+$ sudo apt-get install git
+```
+
+* Install curl
````
`curl` and the Habitat CLI are both installed by `provision.sh` (indirectly), so those steps can be omitted.

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
+* Run the provision script
+
+```
+$ sudo ./support/linux/provision.sh
````
To be safe I think it's good to run this as
```
$ sudo bash -ex habitat/support/linux/provision.sh
```
That way if something fails the script stops and the `-x` tracing makes it easy to see where it failed.

baumanj replied:
Once https://github.com/habitat-sh/habitat/pull/4315 is merged, we should be able to run this command without `sudo`.

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
+
+* After the script completes, source cargo/rust
+```
+$ source $HOME/.cargo/env
````
It may be safer/easier to `source $HOME/.profile` since `provision.sh` may add other things as well. Plus, one would want to do this sourcing after adding the `HAB_` environment variables to the `.profile` (or `.bashrc`, which `.profile` sources) anyway.

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
+### Builder Setup
+* Update your system
+```
+$ sudo apt-get update
````
It looks like the ubuntu 17.10 server image has an entry in `/etc/apt/sources.list` which causes this command to fail
```
E: The repository 'cdrom://Ubuntu-Server 17.10 _Artful Aardvark_ - Release amd64 (20171017.1) artful Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```
This item is already commented out in the 17.10 desktop ISO, but it may be worth mentioning here that it needs to be deleted commented out if using the server image. This command will accomplish that:
```
$ sudo sed -i.bak '/deb cdrom/s/^/#/g' /etc/apt/sources.list
```

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
-`config_worker.toml`
-```toml
+```
+$ sudo vim /hab/svc/builder-worker
````
I think this is missing a `/config.toml` on the end

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
+```
+$ hab install core/hab-backline
+$ cd /hab/cache/artifacts
+$ hab pkg upload -c stable core-hab-backline...hart
````
I think it'd be useful to mention here that if you get
```
✗✗✗
✗✗✗ No auth token specified
✗✗✗
```
That just means you need to add `HAB_AUTH_TOKEN="<your github token>"` to your environment.

baumanj replied:
And if you get 
```
✗✗✗
✗✗✗ No such file or directory (os error 2)
✗✗✗
```
It means that the public key from `bldr.habitat.sh` (that corresponds to the private key the hart file was signed with) wasn't installed.

baumanj commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,294 @@
+* Hit Ctrl + D twice to close the stream
 
-Note: this only needs to be done the first time to get the pieces in place. If you clean the project, make sure to unset the variables show below when you do a clean build, otherwise the build will look for packages on your local build service!
+### Install Dependencies in your local Builder Env
```
I'm not sure all 4 of these `hab origin key import`s are necessary. Installing the `bldr.habitat.sh` public key definitely is (or else `.hart` files `hab install`ed from `bldr.habitat.sh` couldn't be `hab package upload`ed), but I'm not sure about the other 3.

baumanj replied:
I confirmed that I can successfully build packages with only the one key installed.

chefsalim commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,275 @@
-sessionsrv: target/debug/bldr-sessionsrv start --config /home/your_alias/habitat/config_sessionsrv.toml
-worker: target/debug/bldr-worker start --config /home/your_alias/habitat/config_worker.toml
-jobsrv: target/debug/bldr-jobsrv start --config /home/your_alias/habitat/config_jobsrv.toml
+$ sudo apt-get install npm
```
So we should either point to the Web UI README (or pull in some of the best practices from it, like using nvm) here.  The Web UI README is meaty enough that it might be best just to update and refer to it...

chefsalim commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,275 @@
+* Start the UI
+
+```
+$ sudo npm start
````
I don't think we need sudo here...

chefsalim commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,275 @@
 
-Note: this only needs to be done the first time to get the pieces in place. If you clean the project, make sure to unset the variables show below when you do a clean build, otherwise the build will look for packages on your local build service!
+* Navigate to http://localhost:3000/#/pkgs
+* Log in
```
There may be a step here to accept the Habitat Builder Dev app in your github account

chefsalim replied:
Also, it's important that they grant permissions carefully- if the permissions to repos that are granted are too restrictive, they will not be able to link plans (aka create projects) on the repos that are not accessible to the githab app later.

baumanj commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,294 @@
-key_dir = "/path/to/bldr-key-pair"
+### Pre-Reqs
+
+* Ubuntu 17.10 Desktop Virtual Machine
```
Server also works with the one caveat about fixing `/etc/apt/sources.list` below.

baumanj commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,294 @@
 
-    ```
-    HTTP/1.1 201 Created
+## General build notes
````
Is is worth adding anything about the ports that need to be forwarded and how to achieve that to access the web UI from outside the VM?

nellshamrell replied:
We could, but I'd rather they be added after this pull request is resolved/merged. Good to keep in mind!

chefsalim commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,275 @@
+* Open up the builder-worker file
+
+```
+$ sudo vim /hab/svc/builder-worker
````
Does this need a sudo?
Also may be better to specify the config.toml rather than the folder?

baumanj replied:
It does need sudo since typically everything under `/hab` is owned by root.

chefsalim commented on BUILDER_DEV.md:
````diff
@@ -1,390 +1,275 @@
+$ cd /hab/cache/artifacts
+$ hab pkg upload -c stable core-hab-backline...hart
+```
+* You will need to follow this same process for any dependencies of anything you want to build locally - the sample app I will use below depends on [core/scaffolding-node](https://bldr.habitat.sh/#/pkgs/core/scaffolding-node/latest), [core/git](https://bldr.habitat.sh/#/pkgs/core/git/latest), [core/node](https://bldr.habitat.sh/#/pkgs/core/node/latest), [core/busybox-static](https://bldr.habitat. ...
````
Minor - may want to switch back out of first person to be consistent with the rest of the doc.

chefsalim commented on BUILDER_DEV.md:
```diff
@@ -1,390 +1,275 @@
-1. Export the following environment variables:
+### Install Dependencies in your local Builder Env
+
+* on your VM, go to the production instance of Builder (the one at habitat.sh)
```
Minor - Caps for 'on'

baumanj commented:
@thesentinels approve

thesentinels commented:
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

baumanj commented:
![keepgoing](https://user-images.githubusercontent.com/5906042/34134036-4187f32c-e426-11e7-94d4-c84a49b62d21.png)

thesentinels commented:
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/318365034).

thesentinels commented:
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/318365034).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

![gif-keyboard-3280869874741411265](https://cloud.githubusercontent.com/assets/4304/17755674/1e577b82-6490-11e6-96b7-1663d283c824.gif)
